sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1053,6 +1055,41 @@ object PruneFilters extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+ * Collapse plans consisting all empty local relations generated by [[PruneFilters]].
+ * Note that the ObjectProducer/Consumer and direct aggregations are the exceptions.
+ * {{{
+ *   SELECT a, b FROM t WHERE 1=0 GROUP BY a, b ORDER BY a, b ==> empty result
+ *   SELECT SUM(a) FROM t WHERE 1=0 GROUP BY a HAVING COUNT(*)>1 ORDER BY a (Not optimized)
+ * }}}
+ */
+object CollapseEmptyPlan extends Rule[LogicalPlan] with PredicateHelper {
+  private def isEmptyLocalRelation(plan: LogicalPlan): Boolean =
+    plan.isInstanceOf[LocalRelation] && plan.asInstanceOf[LocalRelation].data.isEmpty
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case x if x.isInstanceOf[ObjectProducer] || x.isInstanceOf[ObjectConsumer] => x
+
+    // Case 1: If groupingExpressions contains all aggregation expressions, the result is empty.
+    case a @ Aggregate(ge, ae, child) if isEmptyLocalRelation(child) && ae.forall(ge.contains(_)) =>
+      LocalRelation(a.output, data = Seq.empty)
+
+    // Case 2: General aggregations can generate non-empty results.
+    case a: Aggregate => a
+
+    // Case 3: The following non-leaf plans having only empty relations return empty results.
+    case p: LogicalPlan if p.children.nonEmpty && p.children.forall(isEmptyLocalRelation) =>
+      p match {
+        case _: Project | _: Generate | _: Filter | _: Sample | _: Join |
```

Comment:
actually for intersect you only need one child to be empty

for join if it is inner join you just need one child to be empty too